File: frankpiv/backend/frankr_backend.py
from abc import ABC

from cfrankr import Robot, MotionData, Affine

from frankpiv.backend.general import GeneralBackend


class Backend(GeneralBackend, ABC):

    def __init__(self, config):
        super().__init__(config)
        self.frankr_config = config["frankr"]
        self._robot = None
        self._motion_data = None

    def _initialize(self):
        self._robot = Robot(self.frankr_config["robot_name"], self.frankr_config["dynamic_rel"])
        self._motion_data = MotionData()

    def _finish(self):
        self._robot = None
        self._motion_data = None

    def _current_pose(self) -> Affine:
        return self._robot.current_pose(Affine())

    def _move_robot_cartesian(self, target_pose: Affine):
        self._robot.move_cartesian(Affine(), target_pose, self._motion_data)

File: src/frankr_backend.cpp
#include <utility>

#include "frankpiv/frankr_backend.hpp"
#include "frankpiv/utilities.hpp"

using namespace Eigen;
using namespace frankpiv::util;

namespace frankpiv::backend {
    FrankrBackend::FrankrBackend(const YAML::Node &config, std::string node_name) : GeneralBackend(config, std::move(
            node_name)) {
        YAML::Node frankx_config = config["frankr"];
        this->robot_name = get_config_value<std::string>(frankx_config, "robot_name")[0];
        this->dynamic_rel = get_config_value<double>(frankx_config, "dynamic_rel")[0];
        this->robot = nullptr;
        this->motion_data = nullptr;
    }

    FrankrBackend::~FrankrBackend() {
        this->stop();
    }

    void FrankrBackend::initialize() {
        this->robot = new Robot(this->robot_name, this->dynamic_rel);
        MotionData _motion_data;
        this->motion_data = &_motion_data;
    }

    void FrankrBackend::finish() {
        this->robot = nullptr;
        this->motion_data = nullptr;
    }

    Affine3d FrankrBackend::currentPose() {
        return this->robot->currentPose(Affine()).data;
    }

    void FrankrBackend::moveRobotCartesian(const Affine3d &target_pose) {
        this->robot->moveCartesian(Affine(), target_pose, *this->motion_data);
    }
}